src/components/PastCommitNode.tsx

```diff
@@ -1,114 +1,139 @@
 import {
   pastCommitNodeStyle,
-  pastCommitWorkingNodeStyle,
-  pastCommitContentStyle,
-  pastCommitWorkingContentStyle,
-  pastCommitHeadContentStyle,
-  pastCommitNumberContentStyle,
-  pastCommitActiveContentStyle,
-  pastCommitLineStyle
-} from '../componentsStyle/PastCommitNodeStyle';
+  pastCommitHeaderStyle,
+  pastCommitHeaderItemStyle,
+  pastCommitBodyStyle,
+  branchStyle,
+  branchesStyle,
+  remoteBranchStyle,
+  localBranchStyle,
+  workingBranchStyle,
+  pastCommitExpandedStyle,
+  collapseStyle
+} from "../componentsStyle/PastCommitNodeStyle";
 
-import { classes } from 'typestyle';
+import { classes } from "typestyle";
 
-import * as React from 'react';
+import { SingleCommitInfo, GitBranchResult } from "../git";
+
+import { SinglePastCommitInfo } from "./SinglePastCommitInfo";
+
+import * as React from "react";
 
 export interface IPastCommitNodeProps {
-  index: number;
-  isLast: boolean;
-  pastCommit: any;
-  currentFileBrowserPath: string;
-  setShowList: Function;
-  getPastCommit: Function;
-  activeNode: number;
-  updateActiveNode: Function;
-  isVisible: boolean;
+  pastCommit: SingleCommitInfo;
+  data: GitBranchResult["branches"];
+  topRepoPath: string;
+  currentTheme: string;
+  app: any;
+  diff: any;
+  refresh: any;
+}
+
+export interface IPastCommitNodeState {
+  expanded: boolean;
 }
 
-export class PastCommitNode extends React.Component<IPastCommitNodeProps, {}> {
+export class PastCommitNode extends React.Component<
+  IPastCommitNodeProps,
+  IPastCommitNodeState
+> {
   constructor(props: IPastCommitNodeProps) {
     super(props);
+    this.state = {
+      expanded: false
+    };
   }
-
-  getPastCommitNodeClass(): string {
-    if (!this.props.isVisible) {
-      return null;
+  getBranchesForCommit() {
+    const idAbrev = this.props.pastCommit.commit.slice(0, 7);
+    const branches = [];
+    for (let i = 0; i < this.props.data.length; i++) {
+      const branch = this.props.data[i];
+      // tag sent from describe command. Must unparse to find commit hash
+      // https://git-scm.com/docs/git-describe#git-describe-ltcommit-ishgt82308203
+      if (!branch.tag) {
+        continue;
+      }
+      const tagParts = branch.tag.split("-");
```

Comment:
Do you think it would be more appropriate to return the top commit SHA1 for the branch in the branch API payload instead of determining it from the tag?

In which case the top_commit_sha would always be defined for a given branch.

```
"branches": [
    {
        "is_current_branch":false,
        "is_remote_branch":true,
        "name":"branch-name", 
        "upstream":"upstream-branch-name",
        "tag":"branch-tag",
        "top_commit_sha":"abcd123",
    }
]
```

Reply:
Yeah it would be more appropriate. I tried not to modify the backend API in this PR, to minimize breaking changes.

Should I add that field?

Reply:
This PR is already very large, so If you don't mind could you send another PR as follow up for this comment?

Reply:
Will do, opened https://github.com/jupyterlab/jupyterlab-git/issues/274 to track this